lnp2p/peermgr.go

```diff
@@ -156,81 +156,81 @@ func (pm *PeerManager) TryConnectAddress(addr string, settings *NetSettings) (*P
 
 }
 
-func (pm *PeerManager) tryConnectPeer(netaddr string, lnaddr *lncore.LnAddr, settings *NetSettings) (*Peer, error) {
+func (pm *PeerManager) tryConnectPeer(netaddr string, lnaddr *lncore.LnAddr) (*Peer, error) {
 
 	// lnaddr check, to make sure that we do the right thing.
 	if lnaddr == nil {
 		return nil, fmt.Errorf("connection to a peer with unknown lnaddr not supported yet")
 	}
 
-	// Do NAT setup stuff.
-	if settings != nil && settings.NatMode != nil {
-
-		// Do some type juggling.
-		x, err := strconv.Atoi(netaddr[1:])
-		if err != nil {
-			return nil, err
-		}
-		lisPort := uint16(x) // if only Atoi could infer which type we wanted to parse as!
-
-		// Actually figure out what we're going to do.
-		if *settings.NatMode == "upnp" {
-			// Universal Plug-n-Play
-			logging.Infof("Attempting port forwarding via UPnP...")
-			err = nat.SetupUpnp(lisPort)
-			if err != nil {
-				return nil, err
-			}
-		} else if *settings.NatMode == "pmp" {
-			// NAT Port Mapping Protocol
-			timeout := time.Duration(10 * time.Second)
-			logging.Infof("Attempting port forwarding via PMP...")
-			_, err = nat.SetupPmp(timeout, lisPort)
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			return nil, fmt.Errorf("invalid NAT type: %s", *settings.NatMode)
-		}
-	}
-
 	dialer := net.Dial
 
 	// Use a proxy server if applicable.
-	if settings != nil && settings.ProxyAddr != nil {
-		d, err := connectToProxyTCP(*settings.ProxyAddr, settings.ProxyAuth)
+	ns := pm.netsettings
+	if ns != nil && ns.ProxyAddr != nil {
+		d, err := connectToProxyTCP(*ns.ProxyAddr, ns.ProxyAuth)
 		if err != nil {
 			return nil, err
 		}
 		dialer = d
 	}
 
-	// Set up the connection.
+	// Create the connection.
 	lndcconn, err := lndc.Dial(pm.idkey, netaddr, string(*lnaddr), dialer)
 	if err != nil {
 		return nil, err
 	}
 
-	pi, err := pm.peerdb.GetPeerInfo(*lnaddr)
+	// Try to set up the new connection.
+	p, err := pm.handleNewConnection(lndcconn, *lnaddr)
 	if err != nil {
-		logging.Errorf("Problem loading peer info from DB: %s\n", err.Error())
-		// don't kill the connection?
+		return nil, err
 	}
 
+	// Now start listening for inbound traffic.
+	// (it *also* took me a while to realize I forgot *this*)
+	go processConnectionInboundTraffic(p, pm)
+
+	// Return
+	return p, nil
+
+}
+
+func (pm *PeerManager) handleNewConnection(conn *lndc.Conn, expectedAddr lncore.LnAddr) (*Peer, error) {
+
 	// Now that we've got the connection, actually create the peer object.
-	pk := pubkey(lndcconn.RemotePub())
+	pk := pubkey(conn.RemotePub())
 	rlitaddr := convertPubkeyToLitAddr(pk)
+
+	if rlitaddr != expectedAddr {
+		conn.Close()
+		return nil, fmt.Errorf("peermgr: Connection init error, expected addr %s got addr %s", expectedAddr, rlitaddr)
+	}
+
 	p := &Peer{
 		lnaddr:   rlitaddr,
 		nickname: nil,
-		conn:     lndcconn,
+		conn:     conn,
 		idpubkey: pk,
+
+		// TEMP
+		idx: nil,
+	}
+
+	pi, err := pm.peerdb.GetPeerInfo(expectedAddr)
+	if err != nil {
```

Comment:
This check could actually be `if err != nil || pi == nil {`

Reply:
Also, if we fall through to `logging.Errorf("Problem getting unique peeridx from DB: %s\n", err.Error())`, could we assign a random number above 10000 or something? Because we don't catch that and it could result in some weird case where a peer doesn't have a peerId.

Reply:
My goal is to eventually not have peer IDs externally and at most use them ephemerally within the process.

I guess that works though?  I don't really see how it solves the problem though.

Reply:
We could do way with peer IDs, but that would remove of a wait to easily reference them by the end user. I guess they can do something similar with nicknames, but we must display that in `lit-af`'s output.

It doesn't solve the problem (which in this case is a peer not having a peerId), but circumvents it by assigning a random peerId. Given that the case where not having a peerId is rare (and maybe never occurs in practice), maybe we could just skip it totally.

Reply:
RPCs for BOLT-family LN imlplementations just use the peer's full address, which imo makes sense.  And since our lnaddrs are shorter it's more readable anyways too.  And nicknames would make aliasing that simpler as well, perhaps we should allow the users to provide a nickname when initially connecting.

Reply:
agreed.